File: TimeCardClientAndroid/app/src/main/java/jp/uhimania/timecardclientandroid/data/source/NetworkDataSource.kt
package jp.uhimania.timecardclientandroid.data.source

import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import jp.uhimania.timecardclientandroid.data.TimeRecord
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import retrofit2.Retrofit
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface NetworkDataSource {
    suspend fun getRecords(year: Int, month: Int): List<TimeRecord>
    suspend fun insertRecord(record: TimeRecord): TimeRecord
    suspend fun updateRecord(record: TimeRecord): TimeRecord
    suspend fun deleteRecord(record: TimeRecord)
}

class RetrofitNetworkDataSource(private val baseUrl: HttpUrl) : NetworkDataSource {
    private val json = Json { ignoreUnknownKeys = true }

    private val retrofit = Retrofit.Builder()
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .baseUrl(baseUrl)
        .build()

    private val service: TimeCardApiService by lazy {
        retrofit.create(TimeCardApiService::class.java)
    }

    override suspend fun getRecords(year: Int, month: Int): List<TimeRecord> {
        return service.getRecords(year, month)
    }

    override suspend fun insertRecord(record: TimeRecord): TimeRecord {
        return service.insertRecord(record)[0]
    }

    override suspend fun updateRecord(record: TimeRecord): TimeRecord {
        return service.updateRecord(record.id, record)[0]
    }

    override suspend fun deleteRecord(record: TimeRecord) {
        service.deleteRecord(record.id)
    }
}

interface TimeCardApiService {
    @GET("timecard/records")
    suspend fun getRecords(
        @Query("year") year: Int,
        @Query("month") month: Int
    ): List<TimeRecord>

    @POST("timecard/records")
    suspend fun insertRecord(
        @Body record: TimeRecord
    ): List<TimeRecord>

    @PATCH("timecard/records/{id}")
    suspend fun updateRecord(
        @Path("id") id: String,
        @Body record: TimeRecord
    ): List<TimeRecord>

    @DELETE("timecard/records/{id}")
    suspend fun deleteRecord(
        @Path("id") id: String
    )
}

File: TimeCardClientIOS/TimeCardClientIOS/Data/Source/NetworkDataSource.swift
//
//  NetworkDataSource.swift
//  TimeCardClientIOS
//
//  Created by uhimania on 2025/12/31.
//

import Foundation

protocol NetworkDataSource {
    func getRecords(year: Int, month: Int) async throws -> [TimeRecord]
    func insertRecord(_ record: TimeRecord) async throws -> TimeRecord
    func updateRecord(_ record: TimeRecord) async throws -> TimeRecord
    func deleteRecord(_ record: TimeRecord) async throws
}

class DefaultNetworkDataSource: NetworkDataSource {
    struct NetworkError: Error {
        let status: Int?
    }
    
    private let baseUrl: URL
    
    init(_ baseUrl: URL) {
        self.baseUrl = baseUrl
    }
    
    func getRecords(year: Int, month: Int) async throws -> [TimeRecord] {
        var url = baseUrl.appending(path: "timecard/records")
        url = url.appending(queryItems: [.init(name: "year", value: String(year))])
        url = url.appending(queryItems: [.init(name: "month", value: String(month))])
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse,
           response.statusCode != 200 {
            throw NetworkError(status: response.statusCode)
        }
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        var records = try decoder.decode([TimeRecord].self, from: data)
        for i in records.indices {
            records[i].breakTimes.sort { $0.start ?? .distantPast < $1.start ?? .distantPast }
        }
        
        return records
    }
    
    func insertRecord(_ record: TimeRecord) async throws -> TimeRecord {
        let url = baseUrl.appending(path: "timecard/records")
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let json = try encoder.encode(record)
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = json
        
        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse,
           response.statusCode != 200 {
            throw NetworkError(status: response.statusCode)
        }
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let records = try decoder.decode([TimeRecord].self, from: data)
        guard let record = records.first else {
            throw NetworkError(status: nil)
        }
        
        return record
    }
    
    func updateRecord(_ record: TimeRecord) async throws -> TimeRecord {
        var url = baseUrl.appending(path: "timecard/records")
        url = url.appending(path: record.id.uuidString)
        
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        let json = try encoder.encode(record)
        
        var request = URLRequest(url: url)
        request.httpMethod = "PATCH"
        request.httpBody = json
        
        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse,
           response.statusCode != 200 {
            throw NetworkError(status: response.statusCode)
        }
        
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let records = try? decoder.decode([TimeRecord].self, from: data)
        guard let record = records?.first else {
            throw NetworkError(status: nil)
        }
        
        return record
    }
    
    func deleteRecord(_ record: TimeRecord) async throws {
        var url = baseUrl.appending(path: "timecard/records")
        url = url.appendingPathComponent(record.id.uuidString)
        
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        
        let (_, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse,
           response.statusCode != 200 {
            throw NetworkError(status: response.statusCode)
        }
    }
}
